### PR TITLE
adds support for propagating grpc client-side cancellations

### PR DESCRIPTION
**Pull request by shamsimam**
**jet changes**: 
- [propagate abort errors in the response](https://github.com/twosigma/jet/pull/33)
- [avoids eagerly closing body channel while response streaming](https://github.com/twosigma/jet/pull/34)

## Changes proposed in this PR

- adds support for grpc request state tracking
- attributes frontend request errors as client-error while processing a request
- adds tests for grpc client deadline exceeded and other client cancellations
- adds delay before requesting for request state for server-side cancellation

## Why are we making these changes?

We should propagate any client-side cancellations to the server so that backend connections can be cleaned up.

### Date race in handling the abort request:

There is a data race in handling the abort request versus closing the request body channel which triggers normal completion if the abort request has not been porcessed. This manifest as the following the courier logs:

No data race (note `endStream=false` in the preceding data frame):
```
(io.grpc.netty.shaded.io.netty.handler.codec.http2.Http2FrameLogger logData) [id: 0x786c9352, L:/172.17.0.7:31916 - R:/172.17.0.1:56044] INBOUND DATA: streamId=9 padding=0 endStream=false length=293 bytes=000...
...   
(io.grpc.netty.shaded.io.netty.handler.codec.http2.Http2FrameLogger logRstStream) [id: 0x786c9352, L:/172.17.0.7:31916 - R:/172.17.0.1:56044] INBOUND RST_STREAM: streamId=9 errorCode=8 

```

Date race (`endStream=true` in preceding data frame before reset frame comes along):
```
(io.grpc.netty.shaded.io.netty.handler.codec.http2.Http2FrameLogger logData) [id: 0xe6b87497, L:/172.17.0.7:31124 - R:/172.17.0.1:58702] INBOUND DATA: streamId=13 padding=0 endStream=true length=0 bytes=...
(io.grpc.netty.shaded.io.netty.handler.codec.http2.Http2FrameLogger logRstStream) [id: 0xe6b87497, L:/172.17.0.7:31124 - R:/172.17.0.1:58702] INBOUND RST_STREAM: streamId=13 errorCode=8  
```

### CODE REVIEW

File: containers/test-apps/courier/src/main/java/com/twosigma/waiter/courier/GrpcClient.java [hunk 1]
```diff
@@ -70,6 +73,54 @@ public Status status() {
...
+                } else {
+                    final StatusRuntimeException error = Status.CANCELLED
+                        .withDescription("Observer: " + message)
+                        .asRuntimeException();
+                    observer.cancel(message, error);
+                }
+            }
+        };
+
+        public abstract void apply(final Context.CancellableContext cancellableContext,
```
Comment (sradack) [inline containers/test-apps/courier/src/main/java/com/twosigma/waiter/courier/GrpcClient.java]: Nitpick: Can we avoid requiring parameters for implementations that don't care about the various arguments?

File: containers/test-apps/courier/src/main/java/com/twosigma/waiter/courier/GrpcServer.java [hunk 2]
```diff
@@ -36,12 +44,81 @@
...
+    private final static Context.Key<CidTimestamp> CID_TIMESTAMP = Context.key("CID.TIMESTAMP");
+    private final static int MAX_STATES_TRACKED = 100;
+    private final static Map<String, Map<Long, List<String>>> requestCidToStateList;
 
-    private static void sleep(final int durationMillis) {
-        try {
-            Thread.sleep(durationMillis);
-        } catch (final Exception ex) {
-            ex.printStackTrace();
+    static {
```
Comment (sradack) [inline containers/test-apps/courier/src/main/java/com/twosigma/waiter/courier/GrpcServer.java]: Nitpick: Could we have just kept every CID and avoided the complexity?

File: waiter/resources/log4j.properties [hunk 3]
```diff
@@ -38,3 +38,6 @@ log4j.appender.SchedulerAppender.File=log/${waiter.logFilePrefix}scheduler.log
 log4j.appender.SchedulerAppender.DatePattern='.'yyyy-MM-dd
 log4j.appender.SchedulerAppender.layout=org.apache.log4j.PatternLayout
 log4j.appender.SchedulerAppender.layout.ConversionPattern=%d{ISO8601} %-5p %c [%t] - [CID] %m%n
+
+log4j.logger.org.apache.http=DEBUG
+log4j.logger.org.eclipse.jetty=DEBUG
```
Comment (sradack) [inline waiter/resources/log4j.properties]: We ned to restore this file before merging.

File: containers/test-apps/courier/src/main/java/com/twosigma/waiter/courier/GrpcServer.java [hunk 4]
```diff
@@ -36,12 +42,67 @@
...
+            }
+        };
+    }
+
+    private static List<String> trackState(final String correlationId, final long timestamp, final String state) {
+        if (correlationId != null) {
+            synchronized (requestCidToStateList) {
+                final List<String> stateList;
+
+                if (requestCidToStateList.containsKey(correlationId)) {
```
[the remark this answers is not in the document]
Comment (shamsimam) [reply containers/test-apps/courier/src/main/java/com/twosigma/waiter/courier/GrpcServer.java]: Done.

File: containers/test-apps/courier/src/main/java/com/twosigma/waiter/courier/GrpcClient.java (hunk 1 above)
Comment (shamsimam) [reply containers/test-apps/courier/src/main/java/com/twosigma/waiter/courier/GrpcClient.java]: Skipped, as if I continue with the enum solution, then will need casting in the individual methods.

File: containers/test-apps/courier/src/main/java/com/twosigma/waiter/courier/GrpcServer.java (hunk 2 above)
Comment (shamsimam) [reply containers/test-apps/courier/src/main/java/com/twosigma/waiter/courier/GrpcServer.java]: Done.

File: waiter/resources/log4j.properties (hunk 3 above)
Comment (shamsimam) [reply waiter/resources/log4j.properties]: Removed.

**Pull request closed without merging**